src/test/fuzz/base_encode_decode.cpp

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) 2019 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <test/fuzz/fuzz.h>
+
+#include <base58.h>
+#include <util/strencodings.h>
+
+#include <cassert>
+#include <cstdint>
+#include <string>
+#include <vector>
+
+void test_one_input(const std::vector<uint8_t>& buffer)
+{
+    const std::string random_encoded_string(buffer.begin(), buffer.end());
+
+    std::vector<unsigned char> decoded;
+    if (DecodeBase58(random_encoded_string, decoded)) {
+        const std::string encoded_string = EncodeBase58(decoded);
+        assert(encoded_string == SanitizeString(encoded_string, SAFE_CHARS_BASE58));
+        const std::string sanitized_random_encoded_string = SanitizeString(random_encoded_string, SAFE_CHARS_BASE58);
```

Comment:
All this does is strip the whitespace? Seems overkill to extend the generic helper `SanitizeString` for this use case.

Reply:
No, `random_encoded_string` can contain all kinds of crap and still pass `DecodeBase58` so it is not only whitespace. To properly fuzz test this knowledge of `SAFE_CHARS_BASE58` is needed, but if you want I can make the sanitation machinery local to the this file?

Reply:
Can you explain what kind of crap makes it pass? And does that sound like a bug where it should return false instead?

Reply:
@MarcoFalke Oh, I don't think it is an indication of a bug -- sorry for using sloppy language :)  With "all kinds of crap" I simply meant characters outside of `SAFE_CHARS_BASE58`. I'll get back with a more specific code example :)

Reply:
The reason for this weirdness is that `DecodeBase58(const std::string& str, …)` calls `DecodeBase58(const char* psz, …)` which will stop at first `\x00`.

Consider these cases:
* `str` contains `good` → `DecodeBase58(const std::string& str, …) == true`
* `str` contains `bad0IOl` → `DecodeBase58(const std::string& str, …) == false`
* `str` contains `good\x00bad0IOl` → `DecodeBase58(const std::string& str, …) == true`

Reply:
So instead of `SanitizeString`, the string could be shortened and trimmed of whitespace in the fuzzer?

Reply:
I'm not sure that is the way to go: `\x00` is not whitespace :)

WDYT: shouldn't `DecodeBase58(const std::string& str, …)` return `false` when passed a `str` which contains at least one character that is not any of the base58 characters (all alphanumeric characters except for "0", "I", "O", and "l") or the whitespace characters?

Reply:
Now adjusted fuzz tested code to make the invariant more sane :)